src/Tribe/Linked_Posts.php

```diff
@@ -0,0 +1,812 @@
+<?php
+/**
+ * Object used to handle the linking/unlinking of post types for events
+ */
+class Tribe__Events__Linked_Posts {
+	/**
+	 * @var string Meta key prefix for linked posts
+	 */
+	const META_KEY_PREFIX = '_tribe_linked_post_';
+
+	/**
+	 * @var Tribe__Events__Linked_Posts Singleton instance of the class
+	 */
+	public static $instance;
+
+	/**
+	 * @var Tribe__Events__Main Singleton
+	 */
+	public $main;
+
+	/**
+	 * @var array Collection of post types that can be linked with events
+	 */
+	public $linked_post_types;
+
+	/**
+	 * Returns a singleton of this class
+	 *
+	 * @return Tribe__Events__Linked_Posts
+	 */
+	public static function instance() {
+		if ( ! self::$instance ) {
+			self::$instance = new self;
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Constructor!
+	 */
+	public function __construct() {
+		$this->main = Tribe__Events__Main::instance();
+		$this->register_default_linked_post_types();
+
+		add_action( 'wp_enqueue_scripts', array( $this, 'enqueue_scripts' ) );
+		add_action( 'admin_enqueue_scripts', array( $this, 'enqueue_scripts' ) );
+	}
+
+	public function enqueue_scripts() {
+		$data = array(
+			'post_types' => array_keys( $this->linked_post_types ),
+		);
+		wp_localize_script( 'jquery', 'tribe_events_linked_posts', $data );
+	}
+
+	/**
+	 * Registers the default linked post types for events
+	 *
+	 * @since 4.2
+	 */
+	public function register_default_linked_post_types() {
+		$default_post_types = array(
+			Tribe__Events__Main::VENUE_POST_TYPE,
+			Tribe__Events__Main::ORGANIZER_POST_TYPE,
+		);
+
+		/**
+		 * Filters the list of default registered linked post types
+		 *
+		 * @since 4.2
+		 *
+		 * @var array Array of post type strings
+		 */
+		$linked_post_types = apply_filters( 'tribe_events_register_default_linked_post_types', $default_post_types );
+
+		foreach ( $linked_post_types as $post_type ) {
+			$this->register_linked_post_type( $post_type );
+		}
+	}
+
+	/**
+	 * Registers a post type as a linked post type for events
+	 *
+	 * Notable arguments that can be passed/filtered while registering linked post types:
+	 * - All of the fields returned by get_post_type_object()
+	 * - allow_multiple (default: true) specifies how many of the post type can be linked with an event
+	 * - allow_creation (default: false) specifies whether or not post creation should be allowed
+	 *
+	 * @since 4.2
+	 *
+	 * @param string $post_type Post type slug
+	 * @param array $args Arguments for the linked post type - note: gets merged with get_post_type_object data
+	 *
+	 * @return boolean
+	 */
+	public function register_linked_post_type( $post_type, $args = array() ) {
+		if ( $this->is_linked_post_type( $post_type ) ) {
+			return false;
+		}
+
+		if ( ! $post_type_object = get_post_type_object( $post_type ) ) {
+			return false;
+		}
+
+		$args = wp_parse_args( $args, tribe_object_to_array( $post_type_object ) );
```

Comment:
Where is `tribe_object_to_array()` defined (is this in common but we don't have a PR for it yet)?